doc/architecture/profiling/README.md

```diff
@@ -0,0 +1,96 @@
+# Profiling
+
+## cProfile
+cProfile is a deterministic profilers in which trace functions are executed at various points of interest (e.g. function call, function return, exceptions) and precise timings of these events are recorded.
+
+To profile the application using cProfile, run `make profile`. 
+Profiling information will be collected per-request in the `profiling` directory where it can be examined using the Pstats Interactive Browser.
+
+To load the file into the interactive browser where it can be sorted and queried as required run:
+```bash
+python -m pstats [filename].prof
+```
+
+The profiles can also be combined to give an overview of the profile between all requests.
+
+Combine all the profiles in the 'profiling' directory using:
+```bash
+pipenv run python scripts/merge_profiles.py
+```
+Ensure you delete all the files in this directory before starting your profiling session.
+This will create a file called `combined_profile.prof`
+
+To visualise this profile, `snakeviz` or `gprof2dot` can be used.
+
+
+### Visualisation
+There are many handy profilers but they lack nice visualization interface. We use `snakeviz` and `gprof2dot` for this.
+
+---
+
+#### SnakeViz
+[SnakeViz](https://jiffyclub.github.io/snakeviz/#snakeviz) is a browser based graphical visualisation tool to display profiles using Icicle and Sunburst plots. 
+It also includes IPython line and cell magics that can help to directly profile single line or code blocks and then visualise them. 
+
+Install using:
+```bash
+pip install snakeviz
+```
+##### Visualise a profile:
+```bash
+snakeviz profiling/[filename].prof
+```
+For example:
+```bash
+snakeviz profiling/GET.questionnaire.31ms.1571053121.prof
+```
+
+---
+
+#### gprof2dot
+Converts profiler data into call graphs. It allows to filter functions based on metrics threshold and color them nicely with hotspots.
+
+First install `graphviz` and `gprof2dot`:
+```bash
+brew install graphviz
+pip install gprof2dot
+```
+##### Visualise a profile:
+```bash
+pipenv run gprof2dot -f pstats profiling/[filename].prof | dot -Tpng -o output.png
+```
+For example:
+```bash
+pipenv run gprof2dot -f pstats profiling/GET.questionnaire.31ms.1571053121.prof | dot -Tpng -o profile.png
+```
+To visualise the combined profile run:
+```bash
+pipenv run gprof2dot -f pstats combined_profile.prof | dot -Tpng -o combined_profile.png
+```
+
+---
+
+## Py-Spy
+[Py-Spy](https://github.com/benfred/py-spy) is a sampling profiler where instead of tracking every event (e.g. function call), application is periodically interrupted and stack snapshots are collected. 
+The function call stack is then analysed to deduce the execution time of different parts of the application. 
+
+Most of Python profilers modify application execution in some way: profiling code is typically run inside target python process which often slow down application execution. 
```

Comment:
- Most of Python profilers -> Deterministic profilers
- inside target python -> inside the target Python
- slow -> slows